src/fast-path.js

```diff
@@ -280,17 +282,27 @@ FPp.needsParens = function(assumeExpressionContext) {
       }
 
     case "BinaryExpression":
-      // TODO https://github.com/prettier/prettier/issues/907 this is not complete
+      const isLeftOfAForStatement = node => {
+        let i = 0;
+        while (node) {
+          let parent = this.getParentNode(i++);
+          if (!parent) {
+            return false;
+          }
+          if (parent.type === "ForStatement" && parent.init === node) {
+            return true;
+          }
```

Comment:
If you add
```js
          if (n.Statement.check(parent)) {
            return false;
          }
```
here, you can avoid walking all the way up the tree for every `in`. Not sure how much it would help.